src/goto-programs/remove_java_nondet.cpp

```diff
@@ -0,0 +1,389 @@
+  return is_nondet_returning_object(function_call);
+}
+
+// TODO Pascal suggested moving the following function to util, but I'm not
+// sure.  If goto_programt has public data members, this implies that any state
+// that these data members takes is valid for the goto_program as a whole,
+// which is obviously not the case - if I insert a node into
+// goto_programt::instructions without updating the iterator positions, this
+// has a good chance of invalidating the program.  It should be *impossible* to
+// place an object (with the strong exception safety guarantee) into an invalid
+// state using its public interface.  Therefore, I suggest that the data
+// members of goto_programt be made private, and that this function becomes a
+// public member function on goto_programt.
+
+/*******************************************************************\
+
+Function: insert_before
+
+  Inputs:
+    target: A list of goto program instructions which will be modified.
+    pos: The instruction before which further instructions should be inserted.
+    instrs: A new list of instructions to insert.
+
+ Outputs: An iterator past the end of the inserted range.
+
+ Purpose: Return whether the instr is an assignment from the variable with
+          identifier identifier. This function is required (as opposed to a
+          standard insert) because the goto program instructions must be
+          updated after insertion to maintain goto program invariants.
+
+\*******************************************************************/
+
+static goto_programt::targett insert_before(
+  goto_programt::instructionst &target,
+  const goto_programt::targett &pos,
+  const goto_programt::instructionst &instrs)
+{
+  // Store the instruction before the newly-inserted range.
+  assert(pos!=std::begin(instrs));
+  const auto prev=std::prev(pos);
+
+  std::map<goto_programt::const_targett, goto_programt::targett> target_map;
+  for(auto it=instrs.begin(), instr_end=instrs.end();
+      it!=instr_end;
+      ++it)
+  {
+    const auto p=target_map[it]=
+      target.insert(pos, goto_programt::instructiont());
+    *p=*it;
+  }
+
+  for(auto i=std::next(prev); i!=pos; ++i)
+  {
+    for(auto &t : i->targets)
+    {
+      const auto new_target=target_map.find(t);
+      assert(new_target!=std::end(target_map));
+      t=new_target->second;
+    }
+  }
+
+  return pos;
+}
+
+/*******************************************************************\
+
+Function: is_symbol_with_id
+
+  Inputs:
+    expr: The expression which may be a symbol.
+    identifier: Some identifier.
+
+ Outputs: True if the expression is a symbol with the specified identifier.
+
+ Purpose: Return whether the expression is a symbol with the specified
+          identifier.
+
+\*******************************************************************/
+
+static bool is_symbol_with_id(const exprt& expr, const dstringt& identifier)
+{
+  return expr.id()==ID_symbol && expr.get(ID_identifier)==identifier;
+}
```

Comment:
As a general rule, don't use get() directly; convert to the appropriate exprt type and use the accessor methods.